cloudinit/distros/debian.py

```diff
@@ -300,11 +300,38 @@ def update_package_sources(self):
     def get_primary_arch(self):
         return util.get_dpkg_architecture()
 
-    def set_keymap(self, layout, model, variant, options):
-        # Let localectl take care of updating /etc/default/keyboard
-        distros.Distro.set_keymap(self, layout, model, variant, options)
-        # Workaround for localectl not applying new settings instantly
+    def set_keymap(self, layout: str, model: str, variant: str, options: str):
+        # localectl is broken on some versions of Debian. See
+        # https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/2030788 and
+        # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1038762
+        #
+        # Instead, write the file directly. According to the keyboard(5) man
+        # page, this file is shared between both X and the console.
+
+        contents = "\n".join(
+            [
+                "# This file was generated by cloud-init",
+                "",
+                f'XKBMODEL="{model}"',
+                f'XKBLAYOUT="{layout}"',
+                f'XKBVARIANT="{variant}"',
+                f'XKBOPTIONS="{options}"',
+                "",
+                'BACKSPACE="guess"',  # This is provided on default installs
+                "",
+            ]
+        )
+        util.write_file(
+            filename="/etc/default/keyboard",
+            content=contents,
+            mode=644,
```

Comment:
Mode here needs to be octal not decimal, this resulted in some very bad behavior on mantic when cloud-init wrote the file with perms 1204 :)
```
root@testlocale-manticdaily:~# localectl status
Could not get properties: Access Denied

root@testlocale-manticdaily:~# ls -l /etc/default/keyboard 
-rw-r----T 1 root root 120 Aug 25 02:31 /etc/default/keyboard
```
More details in [the bug](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2030788/comments/7)